- [QuaC](#quac)
  - [What is QuaC?](#what-is-quac)
    - [QC tools included](#qc-tools-included)
  - [Pipeline installation](#pipeline-installation)
    - [Requirements](#requirements)
    - [Retrieve pipeline source code](#retrieve-pipeline-source-code)
  - [Environment Setup](#environment-setup)
    - [Requirements](#requirements-1)
    - [Create conda environment](#create-conda-environment)
    - [QuaC workflow config file](#quac-workflow-config-file)
      - [Prepare verifybamid datasets for exome analysis](#prepare-verifybamid-datasets-for-exome-analysis)
    - [Testing](#testing)
  - [How to run QuaC](#how-to-run-quac)
    - [Example usage](#example-usage)
  - [Output](#output)
    - [Dummy pedigree file creator](#dummy-pedigree-file-creator)
  - [Contributing](#contributing)
  - [Changelog](#changelog)

# QuaC

🦆🦆 Don't duck that QC thingy 🦆🦆

## What is QuaC?

QuaC is a pipeline, developed using snakemake, that runs several QC tools and summarizes results for WGS/WES samples at
CGDS. It is designed to run after samples are run through [CGDS's small variant caller
pipeline](https://gitlab.rc.uab.edu/center-for-computational-genomics-and-data-science/sciops/pipelines/small_variant_caller_pipeline).


### QC tools included

QuaC quacks using the tools listed below:

| Tool                                                                                                                       | Use                                                                                           |
| -------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------- |
| **BAM QC**                                                                                                                    |                                                                                               |
| [Qualimap](http://qualimap.conesalab.org/)                                                                                 | QCs alignment data in SAM/BAM files                                                           |
| [Picard-CollectMultipleMetrics](https://broadinstitute.github.io/picard/command-line-overview.html#CollectMultipleMetrics) | summarizes alignment metrics from a SAM/BAM file using several modules                        |
| [Picard-CollectWgsMetrics](https://broadinstitute.github.io/picard/command-line-overview.html#CollectWgsMetrics)           | Collects metrics about coverage and performance of whole genome sequencing (WGS) experiments. |
| [mosdepth](https://github.com/brentp/mosdepth)                                                                             | Fast BAM/CRAM depth calculation                                                               |
| [indexcov](https://github.com/brentp/goleft/tree/master/indexcov)                                                          | Estimate coverage from whole-genome bam or cram index                                         |
| [covviz](https://github.com/brwnj/covviz)                                                                                  | Identifies large, coverage-based anomalies                                                    |
| **VCF QC**                                                                                                                    |                                                                                               |
| [bcftools stats](https://samtools.github.io/bcftools/bcftools.html#stats)                                                  | Stats variants in VCF                                                                         |
| **Within-species contamination**                                                                                           |                                                                                               |
| [verifybamid](https://github.com/Griffan/VerifyBamID)                                                                      | Estimates within-species (i.e. cross-sample) contamination                                    |
| **Sex, ancestry and relatedness estimation**                                                                               |                                                                                               |
| [somalier](https://github.com/brentp/somalier)                                                                             | Estimation of sex, ancestry and relatedness                                                   |



**Note:** At CGDS, significant amount of QC is additionally performed as part of the [small variant caller
pipeline](https://gitlab.rc.uab.edu/center-for-computational-genomics-and-data-science/small_variant_caller_pipeline/).
This QuaC pipeline should be treated as a companion to the QC results generated by the small_variant_caller_pipeline.


## Pipeline installation

Installation simply requires fetching the source code. Following are required:

### Requirements

- Git v2.0+
- CGDS GitLab access
- [SSH Key for access](https://docs.uabgrid.uab.edu/wiki/Cheaha_GettingStarted#Logging_in_to_Cheaha) to Cheaha cluster


### Retrieve pipeline source code

Pipeline installation simply requires fetching the source code. This repository use git submodules, which needs to be
pulled when cloning. Go to the directory of your choice and run the command below.

```sh
git clone -b master \
    --recurse-submodules \
    git@gitlab.rc.uab.edu:center-for-computational-genomics-and-data-science/sciops/pipelines/quac.git
```

Note that downloading this repository from GitLab, instead of cloning, may not fetch the submodules included.


## Environment Setup

### Requirements

***Direct***

- [Deep clone of repo](#pipeline-installation) created
- Anaconda/miniconda
    - Tested with Anaconda3/2020.02

***Indirect***

- Snakemake
    - Tested with v6.0.5
    - Gets installed as part of conda environment
- Python
    - Tested with v3.6.3
    - Gets installed as part of conda environment
- slurmpy
    - Tested with v0.0.8
    - Gets installed as part of conda environment
- Singularity
    - Tested with v3.5.2
    - Will be loaded as a module


### Create conda environment

Necessary dependencies for QuaC can installed in a conda environment as shown below:

```sh
module reset
module load Anaconda3/2020.02

# create conda environment. Needed only the first time.
conda env create --file configs/env/quac.yaml
# activate conda environment
conda activate quac

# if you need to update the existing environment
conda env update --file configs/env/quac.yaml
```


### QuaC workflow config file

QuaC requires a workflow config file in yaml format (`configs/workflow.yaml`), which provides filepaths to necessary
dependencies required by certain QC tools. Their format should look like:

```yaml
ref: "path to ref genome path"
somalier:
    sites: "path to somalier's site file"
    labels_1kg: "path to somalier's ancestry-labels-1kg file"
    somalier_1kg: "dirpath to somalier's 1kg-somalier files"
verifyBamID:
    svd_dat_wgs: "path to WGS resources .dat files"
    svd_dat_exome: "path to exome resources .dat files"
```

#### Prepare verifybamid datasets for exome analysis

*This step is necessary only for exome samples.* verifybamid has provided auxiliary resource files, which are necessary
for analysis. However, chromosome contigs do not include `chr` prefix in their exome resource files, which are expected
for our analyses at CGDS. Follow these steps to setup resource files with `chr` prefix in their contig names.

```sh
# cd into exome resources dir
cd <path-to>/VerifyBamID-2.0.1/resource/exome/
sed -e 's/^/chr/' 1000g.phase3.10k.b38.exome.vcf.gz.dat.bed > 1000g.phase3.10k.b38_chr.exome.vcf.gz.dat.bed
sed -e 's/^/chr/' 1000g.phase3.10k.b38.exome.vcf.gz.dat.mu > 1000g.phase3.10k.b38_chr.exome.vcf.gz.dat.mu
cp 1000g.phase3.10k.b38.exome.vcf.gz.dat.UD 1000g.phase3.10k.b38_chr.exome.vcf.gz.dat.UD
cp 1000g.phase3.10k.b38.exome.vcf.gz.dat.V 1000g.phase3.10k.b38_chr.exome.vcf.gz.dat.V
```


### Testing

```sh
python src/run_quac.py \
      --project_name test_project \
      --projects_path .test/ngs-data/ \
      --pedigree .test/configs/project.ped -l -n
```

## How to run QuaC

In the activate conda environment, QuaC is run using script `src/quac.py`. Here are all the options available:

```sh
$ ./src/run_quac.py -h
usage: run_quac.py [-h] [--project_name] [--projects_path] [--pedigree]
                   [--outdir] [-m] [--exome] [--cluster_config] [--log_dir]
                   [-e] [-n] [-l] [--rerun_failed]

Wrapper tool for QuaC pipeline.

optional arguments:
  -h, --help          show this help message and exit

QuaC workflow options:
  --project_name      Project name (default: None)
  --projects_path     Path where all projects are hosted. Dont include
                      project name here. (default:
                      /data/project/worthey_lab/projects/)
  --pedigree          Pedigree filepath. Must correspond to the project
                      supplied via --project_name (default: None)
  --outdir            Out directory path (default:
                      $USER_SCRATCH/tmp/quac/results)
  --exome             Flag to run in exome mode (default: False)

QuaC wrapper options:
  --cluster_config    Cluster config json file. Needed for snakemake to run
                      jobs in cluster. (default: /data/project/worthey_lab/pro
                      jects/experimental_pipelines/mana/quac/configs/cluster_c
                      onfig.json)
  --log_dir           Directory path where logs (both workflow's and
                      wrapper's) will be stored (default:
                      $USER_SCRATCH/tmp/quac/results/../logs)
  -e , --extra_args   Pass additional custom args to snakemake. Equal symbol
                      is needed for assignment as in this example: -e='--
                      forceall' (default: None)
  -n, --dryrun        Flag to dry-run snakemake. Does not execute anything,
                      and just display what would be done. Equivalent to '--
                      extra_args "-n"' (default: False)
  -l, --run_locally   Flag to run the snakemake locally and not as a Slurm
                      job. Useful for testing purposes. (default: False)
  --rerun_failed      Number of times snakemake restarts failed jobs. This may
                      be set to >0 to avoid pipeline failing due to job fails
                      due to random SLURM issues (default: 0)
```


### Example usage

`project_name` and `pedigree` are required options to run the tool.

```sh
# for a WGS projetcquac
python src/run_quac.py --project_name CF_CFF_PFarrell --pedigree data/raw/ped/CF_CFF_PFarrell.ped

# for an exome project
python src/run_quac.py --project_name HCC --pedigree data/raw/ped/HCC.ped --exome

# run for an exome project which is not in the default CGDS projects_path
python src/run_quac.py \
      --project_name UnusualCancers_CMGalluzi \
      --projects_path /data/project/sloss/cgds_path_cmgalluzzi/ \
      --pedigree data/raw/ped/UnusualCancers_CMGalluzi.ped \
      --exome

# for a WGS project, and write results to a dir of choice
python src/run_quac.py \
      --slurm_partition medium \
      --project_name CF_CFF_PFarrell \
      --pedigree data/raw/ped/CF_CFF_PFarrell.ped \
      --outdir /some/lake/with/plenty/ducks/

PROJECT="MuscDyst_SU_MAlexander"
python src/run_quac.py \
      --slurm_partition medium \
      --project_name ${PROJECT} \
      --outdir /data/scratch/manag/tmp/quac/results/test_${PROJECT}/analysis \
      --pedigree data/raw/ped/${PROJECT}.ped -n
```

## Output

QuaC results are stored at path specified via option `--outdir` (default: `$USER_SCRATCH/tmp/quac/results`). This
includes aggregated QC results produced by [multiqc](https://multiqc.info/).

### Dummy pedigree file creator

Script `src/create_dummy_ped.py` creates a "dummy" pedigree file given a project path as input. It's purpose is just to
create a basic pedigree file, which will lack sex (unless project tracking sheet is provided), relatedness and affected
info. See header of the script for usage instructions.

Note that we plan to use phenotips in future to produce fully capable pedigree file. One may manually create them as
well, but this could be error-prone.

## Contributing

If you like to make changes to the source code, please see the [contribution guidelines](./CONTRIBUTING.md).

## Changelog

See [here](./Changelog.md).
