- [QuaC](#quac)
  - [Who am I?](#who-am-i)
  - [Installation](#installation)
  - [Environment Setup](#environment-setup)
    - [Requirements](#requirements)
    - [Setup config file](#setup-config-file)
      - [Prepare verifybamid datasets for exome analysis](#prepare-verifybamid-datasets-for-exome-analysis)
    - [Create conda environment](#create-conda-environment)
    - [Testing](#testing)
  - [How to run QuaC](#how-to-run-quac)
    - [Example usage](#example-usage)
  - [Output](#output)
  - [Contributing](#contributing)
  - [Changelog](#changelog)

# QuaC

 Don't duck that QC thingy 

## Who am I?

QuaC is a pipeline developed using snakemake, which runs a set of selected QC tools on NGS samples. Here are the tools
it can currently quack on:

| Tool                                                              | Use                                                        |
| ----------------------------------------------------------------- | ---------------------------------------------------------- |
| [somalier](https://github.com/brentp/somalier)                    | Estimation of sex, ancestry and relatedness                |
| [verifybamid](https://github.com/Griffan/VerifyBamID)             | Estimates within-species (i.e. cross-sample) contamination |
| [mosdepth](https://github.com/brentp/mosdepth)                    | Fast BAM/CRAM depth calculation                            |
| [indexcov](https://github.com/brentp/goleft/tree/master/indexcov) | Estimate coverage from whole-genome bam or cram index      |
| [covviz](https://github.com/brwnj/covviz)                         | Identifies large, coverage-based anomalies                 |


**Note:** At CGDS, significant amount of QC is additionally performed as part of the [small variant caller
pipeline](https://gitlab.rc.uab.edu/center-for-computational-genomics-and-data-science/small_variant_caller_pipeline/).
This QuaC pipeline should be treated as a companion to the QC results generated by the small_variant_caller_pipeline.


## Installation

Installation simply requires fetching the source code. Following are required:

- Git
- CGDS GitLab access
- [SSH Key for access](https://docs.uabgrid.uab.edu/wiki/Cheaha_GettingStarted#Logging_in_to_Cheaha) to Cheaha cluster

To fetch source code, change in to directory of your choice and run:

```sh
git clone -b master \
    --recurse-submodules \
    git@gitlab.rc.uab.edu:center-for-computational-genomics-and-data-science/sciops/pipelines/quac.git
```

Note that this repository uses git submodules, which gets automatically pulled when cloning using above command. Simply
downloading this repository from GitLab, instead of cloning, may not fetch the submodules included.

## Environment Setup

### Requirements

- Anaconda/miniconda

Also the tools listed below, which are not available via conda distribution, need to be installed. Static binaries are
available for both these tools and they are hence easy to install.

- [somalier](https://github.com/brentp/somalier)
- [goleft](https://github.com/brentp/goleft)

*Note:* CGDS folks using QuaC in cheaha may skip this step, as these tools are already installed and centrally
available.

### Setup config file

Workflow config file `configs/workflow.yaml` provides path to path to necessary QC tools as well as other files that
some QC tools require. Modify them as necessary. Refer to the QC tool's documentation for more information on files that
they require.

#### Prepare verifybamid datasets for exome analysis

*This step is necessary only for exome samples.* verifybamid has provided auxiliary resource files, which are necessary
for analysis. However, chromosome contigs do not include `chr` prefix in their exome resource files, which are expected
for our analysis. Follow these steps to setup resource files with `chr` prefix in their contig names.

```sh
# cd into exome resources dir
cd <path-to>/VerifyBamID-2.0.1/resource/exome/
sed -e 's/^/chr/' 1000g.phase3.10k.b38.exome.vcf.gz.dat.bed > 1000g.phase3.10k.b38_chr.exome.vcf.gz.dat.bed
sed -e 's/^/chr/' 1000g.phase3.10k.b38.exome.vcf.gz.dat.mu > 1000g.phase3.10k.b38_chr.exome.vcf.gz.dat.mu
cp 1000g.phase3.10k.b38.exome.vcf.gz.dat.UD 1000g.phase3.10k.b38_chr.exome.vcf.gz.dat.UD
cp 1000g.phase3.10k.b38.exome.vcf.gz.dat.V 1000g.phase3.10k.b38_chr.exome.vcf.gz.dat.V
```

### Create conda environment

```sh
module reset
module load Anaconda3/2020.02

# create conda environment. Needed only the first time.
conda env create --file configs/env/quac.yaml
# activate conda environment
conda activate quac

# if you need to update the existing environment
conda env update --file configs/env/quac.yaml
```

### Testing

```sh
python src/run_quac.py \
      --project_name test_project \
      --projects_path .test/ngs-data/ \
      --pedigree .test/configs/project.ped
```

## How to run QuaC

In the activate conda environment, QuaC is run using script `src/quac.py`. Here are all the options available:

```sh
$ ./src/run_quac.py -h
usage: run_quac.py [-h] [--project_name] [--projects_path] [--pedigree]
                   [--outdir] [-m] [--exome] [--cluster_config] [--log_dir]
                   [-e] [-n] [-l] [--rerun_failed]

Wrapper tool for QuaC pipeline.

optional arguments:
  -h, --help          show this help message and exit

QuaC workflow options:
  --project_name      Project name (default: None)
  --projects_path     Path where all projects are hosted. Dont include
                      project name here. (default:
                      /data/project/worthey_lab/projects/)
  --pedigree          Pedigree filepath. Must correspond to the project
                      supplied via --project_name (default: None)
  --outdir            Out directory path (default:
                      $USER_SCRATCH/tmp/quac/results)
  -m , --modules      Runs only these user-specified modules(s). If >1, use
                      comma as delimiter. See QuaC snakemake script for
                      available options. Useful for development. (default:
                      all)
  --exome             Flag to run in exome mode (default: False)

QuaC wrapper options:
  --cluster_config    Cluster config json file. Needed for snakemake to run
                      jobs in cluster. (default: /data/project/worthey_lab/pro
                      jects/experimental_pipelines/mana/quac/configs/cluster_c
                      onfig.json)
  --log_dir           Directory path where logs (both workflow's and
                      wrapper's) will be stored (default:
                      $USER_SCRATCH/tmp/quac/results/../logs)
  -e , --extra_args   Pass additional custom args to snakemake. Equal symbol
                      is needed for assignment as in this example: -e='--
                      forceall' (default: None)
  -n, --dryrun        Flag to dry-run snakemake. Does not execute anything,
                      and just display what would be done. Equivalent to '--
                      extra_args "-n"' (default: False)
  -l, --run_locally   Flag to run the snakemake locally and not as a Slurm
                      job. Useful for testing purposes. (default: False)
  --rerun_failed      Number of times snakemake restarts failed jobs. This may
                      be set to >0 to avoid pipeline failing due to job fails
                      due to random SLURM issues (default: 0)
```


### Example usage

`project_name` and `pedigree` are required options to run the tool.

```sh
# for a WGS projetcquac
python src/run_quac.py --project_name CF_CFF_PFarrell --pedigree data/raw/ped/CF_CFF_PFarrell.ped

# for an exome project
python src/run_quac.py --project_name HCC --pedigree data/raw/ped/HCC.ped --exome

# run for an exome project which is not in the default CGDS projects_path
python src/run_quac.py --project_name UnusualCancers_CMGalluzi \
      --projects_path /data/project/sloss/cgds_path_cmgalluzzi/ \
      --pedigree data/raw/ped/UnusualCancers_CMGalluzi.ped \
      --exome

# for a WGS project, run specific modules/tools and write results to a dir of choice
python src/run_quac.py \
      --project_name CF_CFF_PFarrell \
      --pedigree data/raw/ped/CF_CFF_PFarrell.ped \
      --modules somalier \
      --outdir /some/lake/with/plenty/ducks/
```

## Output

QuaC results are stored at path specified via option `--outdir` (default: `$USER_SCRATCH/tmp/quac/results`). This
includes aggregated QC results produced by [multiqc](https://multiqc.info/).

## Contributing

If you like to make changes to the source code, please see the [contribution guidelines](./CONTRIBUTING.md).

## Changelog

See [here](./Changelog.md).
